File: my-folder/0394-decode-string/solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        res = ''
        num = ''

        for char in s:
            if char == '[':
                stack.append((res, num))
                res = ''
                num = ''
            elif char == ']':
                prevRes, prevNum = stack.pop()
                res = prevRes + res * int(prevNum)
            elif char.isdigit():
                num += char
            else :
                res += char

        return res

File: my-folder/0394-decode-string/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function(s) {
    const stack = [];
    let result = '';
    let num = '';

    for (const char of s) {
        if (char === '[') {
            stack.push({ result, num });
            result = '';
            num = '';
        } else if (char === ']') {
            const { result: prevResult, num: prevNum } = stack.pop();
            result = prevResult + result.repeat(Number(prevNum));
        } else if (!isNaN(Number(char))) {
            num += char;
        } else {
            result += char;
        }
    }

    return result;
}
